server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java

```diff
@@ -159,32 +158,37 @@ public SearchPhaseController.ReducedQueryPhase reduce() throws Exception {
         if (f != null) {
             throw f;
         }
-
+        List<QuerySearchResult> buffer;
+        synchronized (this) {
+            buffer = this.buffer;
+            buffer = buffer == null ? Collections.emptyList() : buffer;
+            this.buffer = null;
+        }
         // ensure consistent ordering
-        sortBuffer();
+        buffer.sort(RESULT_COMPARATOR);
         final TopDocsStats topDocsStats = this.topDocsStats;
+        var mergeResult = this.mergeResult;
+        this.mergeResult = null;
```

Comment:
Do I understand correctly that this is the core of the fix, to null out the mergeResult reference? I see we do that onMergeFailure, but we failed to do so on reduce? 

I see existing efforts made to null out aggs in releaseAggs, is my understanding correct that that releasing does not affect the reducedAggs within MergeResult, but rather the individual aggs results from not yet merged query shard results?

Reply:
Yes you are correct there, the intermediary merge result was holding on to the agg still and could become very large if the agg didn't reduce well in the partial merge step.

Reply:
thanks for confirming!